third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,305 @@
+  auto factorizedNumWarps = factorizePowerOf2(numWarps);
+  // Create a list of temporary layouts
+  SmallVector<Attribute> tmpLayouts;
+  for (int i = 0; i < factorizedNumWarps.size(); i++) {
+    auto warpsPerCTA = factorizedNumWarps[i];
+    tmpLayouts.push_back(createTmpLayout(srcEnc, warpsPerCTA));
+    tmpLayouts.push_back(createTmpLayout(dstEnc, warpsPerCTA));
+    SmallVector<unsigned> elemsPerThread(rank, 1);
+    SmallVector<unsigned> threadsPerWarp(rank, 1);
+    threadsPerWarp[rank - 1] = warpSize / 8;
+    threadsPerWarp[rank - 2] = warpSize / threadsPerWarp[rank - 1];
+    auto order = triton::gpu::getOrder(srcEnc);
+    auto layoutCTA = triton::gpu::getCTALayout(srcEnc);
+    auto fallbackLayout = triton::gpu::BlockedEncodingAttr::get(
+        ctx, elemsPerThread, threadsPerWarp, warpsPerCTA, order, layoutCTA);
+    tmpLayouts.push_back(fallbackLayout);
+  }
+
+  unsigned minLDSUsage = 2 * LDSSize;
+  int minIdx = -1;
+  for (int i = 0; i < tmpLayouts.size(); i++) {
+    auto tmpLayout = tmpLayouts[i];
+    std::tie(tmpCvt, newEpilogueCvt) =
+        createNewConvertOps(builder, cvtOp, tmpLayout);
+
+    int tmpCvtLDS = getCvtOpLDSUsage(tmpCvt);
+    int newCvtLDS = getCvtOpLDSUsage(newEpilogueCvt);
+    if (tmpCvtLDS <= LDSSize && newCvtLDS <= LDSSize) {
+      int LDSUsage = std::max(tmpCvtLDS, newCvtLDS);
+      if (LDSUsage < minLDSUsage) {
+        minLDSUsage = LDSUsage;
+        minIdx = i;
+      }
+    }
+    newEpilogueCvt.erase();
+    tmpCvt.erase();
+  }
+
+  if (minIdx == -1) {
+    return;
+  }
+
+  assert(minIdx >= 0 && minIdx < tmpLayouts.size());
+  auto tmpLayout = tmpLayouts[minIdx];
+  std::tie(tmpCvt, newEpilogueCvt) =
+      createNewConvertOps(builder, cvtOp, tmpLayout);
+
+  cvtOp.replaceAllUsesWith(newEpilogueCvt.getResult());
+  cvtOp.erase();
+}
+
+/**
+ * @brief Get live information for LDS buffers in function
+ * @return mapping from operation to list of live LDS buffers
+ */
+std::map<mlir::Operation *, SmallVector<Allocation::BufferId>>
+analyzeBufferLiveness(FunctionOpInterface func, const Allocation *allocations) {
+  std::map<mlir::Operation *, SmallVector<Allocation::BufferId>> liveBuffers;
+
+  mlir::Liveness liveness(func);
+  auto analyzeOperation = [&](mlir::Operation *op) -> void {
+    auto scratchBuffer = allocations->getBufferId(op);
+    if (scratchBuffer != Allocation::InvalidBufferId)
+      liveBuffers[op].push_back(scratchBuffer);
+    for (auto result : op->getOpResults()) {
+      auto bufferId = allocations->getBufferId(result);
+      if (bufferId == Allocation::InvalidBufferId)
+        continue;
+      auto liveOperations = liveness.resolveLiveness(result);
+      for (auto depOp : liveOperations)
+        liveBuffers[depOp].push_back(bufferId);
+    }
+  };
+  func.walk(analyzeOperation);
+  return liveBuffers;
+}
+
+SmallVector<triton::gpu::ConvertLayoutOp>
+findLDSBottleneck(ModuleAllocation &allocAnalysis, FunctionOpInterface func) {
```

Comment:
We can also put this to the common part since it can benefit NV path. But after realizing NV GPUs have pretty large shared memory ....